aiobafi6/device.py

```diff
@@ -410,22 +412,38 @@ def _start(self):
         connection, send periodic property queries, and service query commits.
         """
         assert self._run_fut is None
-        assert not self._stop_requested
         _LOGGER.debug("%s: Starting.", self.name)
         self._run_fut = self._loop.create_future()
-        self._run_fut.add_done_callback(self._finish_run)
-        self._sched_connect_or_signal_run_fut()
+
+        def stop_on_done(_: asyncio.Future):
+            self._stop()
+
+        self._run_fut.add_done_callback(stop_on_done)
+        self._sched_connect_or_reset()
 
     def _stop(self) -> None:
-        """Stop the device."""
-        if self._stop_requested:
-            return
+        """Stop the device.
+
+        This function ultimately causes `_sched_connect_or_reset` to be called by
+        cancelling the appropriate in-flight task.
+
+        This function also creates a new available future, thus marking the device as
+        unavailable. If the run future became done because of an exception or because
+        it was cancelled, that is propagated to the prior available future. Otherwise,
+        the prior available future is never signalled."""
         _LOGGER.debug("%s: Stopping.", self.name)
-        # This will cause `_sched_connect_or_signal_run_fut` to signal `_run_fut`.
-        self._stop_requested = True
-        # The device is not available anymore. Dispatch device callbacks so clients can
-        # react to the change.
-        self._available_event.clear()
+        # Propagate run exception or cancellation to the available future, then reset it
+        # to set the device as unavailable.
+        assert self._run_fut
+        if self._run_fut.cancelled():
```

Comment:
On mobile so will need to check it later but if we ever cancel this future ourselves, cancelation could leak into the integration.  I don't think we do though

Reply:
looks safe

Reply:
Unless there is a compelling reason to later refactor this, I think its fine as-is.